test/metabase/driver/postgres_test.clj

```diff
@@ -1697,3 +1698,57 @@
                                                                         nil))
                           (lib/limit 1))]
             (is (->> query qp/process-query mt/rows))))))))
+
+(deftest column-attributes-test
+  (mt/test-driver :postgres
+    (drop-if-exists-and-create-db! "generated_cols_test")
+    (let [details  (mt/dbdef->connection-details :postgres :db {:database-name "generated_cols_test"})
+          spec     (sql-jdbc.conn/connection-details->spec :postgres details)
+          describe
+          (fn [column-defs]
+            (mt/with-temp [:model/Database database {:engine :postgres, :details details}]
+              (jdbc/execute! spec ["DROP TABLE IF EXISTS test_table"])
+              (jdbc/execute! spec [(->> (map #(format "col%d %s" %1 %2) (range) column-defs)
+                                        (str/join ",")
+                                        (format "CREATE TABLE test_table (%s)"))])
+              (->> (driver/describe-fields :postgres database {:table-names ["test_table"]})
+                   (mapv #(-> (select-keys % [:database-is-generated :database-is-nullable :database-default])
+                            ;; shorthand
+                              (set/rename-keys {:database-is-generated :g
+                                                :database-is-nullable  :n
+                                                :database-default      :d}))))))
+          do-test
+          (fn [& test-cases]
+            (let [test-cases' (partition 2 test-cases)
+                  results (describe (map first test-cases'))]
+              (is (= (count test-cases) (* 2 (count results))))
+              (doseq [[[col-def expected] result] (map vector test-cases' results)]
+                (testing col-def
+                  (if (set? expected)
+                    (is (contains? expected result))
+                    (is (= expected result)))))))]
+      (do-test
+       ;; we might want to later distinguish between user specified
+       ;; defaulting and those expressions introduced by SERIAL & BIGSERIAL
+       ;; but for now:
+       "SERIAL"
+       {:g false, :n false, :d "nextval('test_table_col0_seq'::regclass)"}
+
+       "INT"
+       {:g false, :n true}
+
+       "INT GENERATED ALWAYS AS (col0 + 1) STORED"
+       {:g true,  :n true}
+
+       ;; note how postgres is permitted to rewrite the expression
+       "INT DEFAULT EXTRACT(EPOCH FROM NOW())"
+       ;; some versions of postgres only change the casing (e.g postgres:17)
+       #{{:g false, :n true, :d "EXTRACT(epoch FROM now())"}
+         ;; others completely rewrite the expression (e.g postgres:12)
```

Comment:
I hope there isn't more fun like this whenever we bump the versions we test with

Reply:
indeed. Would you just use a more straighforward test here then? I thought the rewriting was kind of interesting so I kept it, once I realised the breakage with versions I did not remove it: Perhaps I should change it to something that is not gonna get rewritten and leave a comment about rewrites somewhere.

Might be less hassle in practice to just ignore the rewrites, not sure if its anything more than trivia.

Reply:
I like this test as it highlights the shaky ground we're on